In [154]:
import pandas as pd

In [233]:
df=pd.read_excel("Data_Train.xlsx")

FLIGHT PRICE PREDICTION : PREDICTING FLIGHT PRICES BASED ON PREVIOUS JOURNEYS

DATASET : 10,000 OBSERVATIONS BASED ON 11 COLUMNS

OBJECTIVE : 

TO PREDICT FLIGHT PRICE BASED ON CUSTOMER NEEDS 

In [234]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [236]:
df.shape

(10683, 11)

In [237]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
Airline            10683 non-null object
Date_of_Journey    10683 non-null object
Source             10683 non-null object
Destination        10683 non-null object
Route              10682 non-null object
Dep_Time           10683 non-null object
Arrival_Time       10683 non-null object
Duration           10683 non-null object
Total_Stops        10682 non-null object
Additional_Info    10683 non-null object
Price              10683 non-null int64
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [238]:
df.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

In [239]:
df.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

# We can see that few dimensions which are of object types need to be converted into numeric/continuous.
-> Airline , Date_of_Journey , Source , Destination , Route , Dep_Time , Arrival_Time , Duration , Total_Stops

# DATA PREPARATION

In [240]:
k=df["Date_of_Journey"].str.split("/",expand=True).astype(int)

In [241]:
k

0  1     2
0      24  3  2019
1       1  5  2019
2       9  6  2019
3      12  5  2019
4       1  3  2019
...    .. ..   ...
10678   9  4  2019
10679  27  4  2019
10680  27  4  2019
10681   1  3  2019
10682   9  5  2019

[10683 rows x 3 columns]

In [242]:
df["month"]=k[1]
df["day"]=k[0]

In [243]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  month  \
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897      3   
1    05:50         13:15   7h 25m     2 stops         No info   7662      5   
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882      6   
3    18:05         23:30   5h 25m      1 stop         No info   6218      5   
4    16:50         21:35   4h 45m      1 stop         No info  13302      3   

   day  
0   24  
1    1  
2    9  
3   12  
4    1

In [244]:
df=df.drop("Date_of_Journey",axis=1)

In [245]:
df.head()

Airline    Source Destination                  Route Dep_Time  \
0       IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL    16:50   

   Arrival_Time Duration Total_Stops Additional_Info  Price  month  day  
0  01:10 22 Mar   2h 50m    non-stop         No info   3897      3   24  
1         13:15   7h 25m     2 stops         No info   7662      5    1  
2  04:25 10 Jun      19h     2 stops         No info  13882      6    9  
3         23:30   5h 25m      1 stop         No info   6218      5   12  
4         21:35   4h 45m      1 stop         No info  13302      3    1

In [246]:
df["Total_Stops"].value_counts()

1 stop      5625
non-stop    3491
2 stops     1520
3 stops       45
4 stops        1
Name: Total_Stops, dtype: int64

In [247]:
df["Total_Stops"]=df["Total_Stops"].replace("non-stop","0")
df["Total_Stops"]=df["Total_Stops"].str.replace("stops","")
df["Total_Stops"]=df["Total_Stops"].str.replace("stop","")

In [248]:
df["Total_Stops"].value_counts()

1     5625
0     3491
2     1520
3       45
4        1
Name: Total_Stops, dtype: int64

In [249]:
df["Total_Stops"]=pd.to_numeric(df["Total_Stops"])

In [250]:
df.dtypes

Airline             object
Source              object
Destination         object
Route               object
Dep_Time            object
Arrival_Time        object
Duration            object
Total_Stops        float64
Additional_Info     object
Price                int64
month                int32
day                  int32
dtype: object

In [251]:
df["Additional_Info"].value_counts()

No info                         8345
In-flight meal not included     1982
No check-in baggage included     320
1 Long layover                    19
Change airports                    7
Business class                     4
No Info                            3
2 Long layover                     1
Red-eye flight                     1
1 Short layover                    1
Name: Additional_Info, dtype: int64

In [252]:
df=df.drop("Additional_Info",axis=1)

In [253]:
df.head()

Airline    Source Destination                  Route Dep_Time  \
0       IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL    16:50   

   Arrival_Time Duration  Total_Stops  Price  month  day  
0  01:10 22 Mar   2h 50m          0.0   3897      3   24  
1         13:15   7h 25m          2.0   7662      5    1  
2  04:25 10 Jun      19h          2.0  13882      6    9  
3         23:30   5h 25m          1.0   6218      5   12  
4         21:35   4h 45m          1.0  13302      3    1

In [254]:
df.shape

(10683, 11)

In [255]:
df["Dep_Time"].value_counts()

18:55    233
17:00    227
07:05    205
10:00    203
07:10    202
        ... 
03:00      1
16:25      1
12:45      1
04:50      1
04:15      1
Name: Dep_Time, Length: 222, dtype: int64

In [256]:
df["Arrival_Time"].value_counts()

19:00           423
21:00           360
19:15           333
16:10           154
12:35           122
               ... 
03:35 25 Jun      1
00:10 22 Mar      1
00:25 19 Mar      1
02:20 04 May      1
00:50 16 May      1
Name: Arrival_Time, Length: 1343, dtype: int64

In [257]:
df["Duration"].value_counts()

2h 50m     550
1h 30m     386
2h 45m     337
2h 55m     337
2h 35m     329
          ... 
29h 40m      1
29h 30m      1
3h 25m       1
32h 20m      1
35h 20m      1
Name: Duration, Length: 368, dtype: int64

In [258]:
df.dtypes

Airline          object
Source           object
Destination      object
Route            object
Dep_Time         object
Arrival_Time     object
Duration         object
Total_Stops     float64
Price             int64
month             int32
day               int32
dtype: object

In [259]:
cat_cols=["Airline","Source","Destination"]

In [260]:
from sklearn import preprocessing

In [261]:
le = preprocessing.LabelEncoder() 
  
# Encode labels in column 'species'. 

df_category=  df[cat_cols].apply(le.fit_transform)


In [262]:
df_category.head()

Airline  Source  Destination
0        3       0            5
1        1       3            0
2        4       2            1
3        3       3            0
4        3       0            5

In [263]:
df=df.drop(cat_cols,axis=1)

In [264]:
df=pd.concat([df,df_category],axis=1)

In [265]:
df.head()

Route Dep_Time  Arrival_Time Duration  Total_Stops  Price  \
0              BLR → DEL    22:20  01:10 22 Mar   2h 50m          0.0   3897   
1  CCU → IXR → BBI → BLR    05:50         13:15   7h 25m          2.0   7662   
2  DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h          2.0  13882   
3        CCU → NAG → BLR    18:05         23:30   5h 25m          1.0   6218   
4        BLR → NAG → DEL    16:50         21:35   4h 45m          1.0  13302   

   month  day  Airline  Source  Destination  
0      3   24        3       0            5  
1      5    1        1       3            0  
2      6    9        4       2            1  
3      5   12        3       3            0  
4      3    1        3       0            5

In [266]:
n=df["Dep_Time"].str.split(":",expand=True).astype(int)

In [267]:
df["Dep_Time_Hours"]=n[0]
df["Dep_Time_Minutes"]=n[1]

In [268]:
df=df.drop("Dep_Time",axis=1)

In [269]:
df.head()

Route  Arrival_Time Duration  Total_Stops  Price  month  \
0              BLR → DEL  01:10 22 Mar   2h 50m          0.0   3897      3   
1  CCU → IXR → BBI → BLR         13:15   7h 25m          2.0   7662      5   
2  DEL → LKO → BOM → COK  04:25 10 Jun      19h          2.0  13882      6   
3        CCU → NAG → BLR         23:30   5h 25m          1.0   6218      5   
4        BLR → NAG → DEL         21:35   4h 45m          1.0  13302      3   

   day  Airline  Source  Destination  Dep_Time_Hours  Dep_Time_Minutes  
0   24        3       0            5              22                20  
1    1        1       3            0               5                50  
2    9        4       2            1               9                25  
3   12        3       3            0              18                 5  
4    1        3       0            5              16                50

In [270]:
m=df["Arrival_Time"].str.split(" ",expand=True)

In [271]:
m.head()

0     1     2
0  01:10    22   Mar
1  13:15  None  None
2  04:25    10   Jun
3  23:30  None  None
4  21:35  None  None

In [272]:
l=m[0].str.split(":",expand=True).astype(int)

In [273]:
l.head()

0   1
0   1  10
1  13  15
2   4  25
3  23  30
4  21  35

In [274]:
df["Arrival_Time_hours"]=l[0]
df["Arrival_Time_minutes"]=l[1]
df["Arrival_Time_day"]=m[1]
df["Arrival_Time_month"]=m[2]

In [275]:
df=df.drop("Arrival_Time",axis=1)

In [276]:
df.head()

Route Duration  Total_Stops  Price  month  day  Airline  \
0              BLR → DEL   2h 50m          0.0   3897      3   24        3   
1  CCU → IXR → BBI → BLR   7h 25m          2.0   7662      5    1        1   
2  DEL → LKO → BOM → COK      19h          2.0  13882      6    9        4   
3        CCU → NAG → BLR   5h 25m          1.0   6218      5   12        3   
4        BLR → NAG → DEL   4h 45m          1.0  13302      3    1        3   

   Source  Destination  Dep_Time_Hours  Dep_Time_Minutes  Arrival_Time_hours  \
0       0            5              22                20                   1   
1       3            0               5                50                  13   
2       2            1               9                25                   4   
3       3            0              18                 5                  23   
4       0            5              16                50                  21   

   Arrival_Time_minutes Arrival_Time_day Arrival_Time_month  
0                    10               22                Mar  
1                    15             None               None  
2                    25               10                Jun  
3                    30             None               None  
4                    35             None               None

In [277]:
df["Duration"].value_counts()

2h 50m     550
1h 30m     386
2h 45m     337
2h 55m     337
2h 35m     329
          ... 
29h 40m      1
29h 30m      1
3h 25m       1
32h 20m      1
35h 20m      1
Name: Duration, Length: 368, dtype: int64

In [278]:
df["Duration"]=df["Duration"].str.replace("h","")
df["Duration"]=df["Duration"].str.replace("m","")

In [279]:
df["Duration"].value_counts()

2 50     550
1 30     386
2 45     337
2 55     337
2 35     329
        ... 
28 30      1
33 20      1
30 10      1
27 30      1
42 45      1
Name: Duration, Length: 367, dtype: int64

In [280]:
p=df["Duration"].str.split(" ",expand=True)

In [281]:
p.head()

0     1
0   2    50
1   7    25
2  19  None
3   5    25
4   4    45

In [282]:
df["Duration_hours"]=p[0]
df["Duration_minutes"]=p[1]

In [283]:
df.head()

Route Duration  Total_Stops  Price  month  day  Airline  \
0              BLR → DEL     2 50          0.0   3897      3   24        3   
1  CCU → IXR → BBI → BLR     7 25          2.0   7662      5    1        1   
2  DEL → LKO → BOM → COK       19          2.0  13882      6    9        4   
3        CCU → NAG → BLR     5 25          1.0   6218      5   12        3   
4        BLR → NAG → DEL     4 45          1.0  13302      3    1        3   

   Source  Destination  Dep_Time_Hours  Dep_Time_Minutes  Arrival_Time_hours  \
0       0            5              22                20                   1   
1       3            0               5                50                  13   
2       2            1               9                25                   4   
3       3            0              18                 5                  23   
4       0            5              16                50                  21   

   Arrival_Time_minutes Arrival_Time_day Arrival_Time_month Duration_hours  \
0                    10               22                Mar              2   
1                    15             None               None              7   
2                    25               10                Jun             19   
3                    30             None               None              5   
4                    35             None               None              4   

  Duration_minutes  
0               50  
1               25  
2             None  
3               25  
4               45

In [284]:
df.dtypes

Route                    object
Duration                 object
Total_Stops             float64
Price                     int64
month                     int32
day                       int32
Airline                   int32
Source                    int32
Destination               int32
Dep_Time_Hours            int32
Dep_Time_Minutes          int32
Arrival_Time_hours        int32
Arrival_Time_minutes      int32
Arrival_Time_day         object
Arrival_Time_month       object
Duration_hours           object
Duration_minutes         object
dtype: object

In [285]:
df["Route"].value_counts()

DEL → BOM → COK                2376
BLR → DEL                      1552
CCU → BOM → BLR                 979
CCU → BLR                       724
BOM → HYD                       621
                               ... 
CCU → IXZ → MAA → BLR             1
CCU → IXA → BLR                   1
BOM → COK → MAA → HYD             1
BLR → HBX → BOM → AMD → DEL       1
BOM → NDC → HYD                   1
Name: Route, Length: 128, dtype: int64

In [286]:
df=df.replace("None",0)

In [287]:
df["Route"]=df["Route"].astype(str)

In [288]:
df["Route"]

0                    BLR → DEL
1        CCU → IXR → BBI → BLR
2        DEL → LKO → BOM → COK
3              CCU → NAG → BLR
4              BLR → NAG → DEL
                 ...          
10678                CCU → BLR
10679                CCU → BLR
10680                BLR → DEL
10681                BLR → DEL
10682    DEL → GOI → BOM → COK
Name: Route, Length: 10683, dtype: object

In [289]:
from sklearn.preprocessing import LabelEncoder 

In [290]:

le = LabelEncoder()

In [291]:
df['Route']=le.fit_transform(df['Route'])

In [292]:
df["Route"].head()

0     18
1     84
2    118
3     91
4     29
Name: Route, dtype: int32

In [293]:
df.dtypes

Route                     int32
Duration                 object
Total_Stops             float64
Price                     int64
month                     int64
day                       int64
Airline                   int64
Source                    int64
Destination               int64
Dep_Time_Hours            int64
Dep_Time_Minutes          int64
Arrival_Time_hours        int64
Arrival_Time_minutes      int64
Arrival_Time_day         object
Arrival_Time_month       object
Duration_hours           object
Duration_minutes         object
dtype: object

In [294]:
df.head()

Route Duration  Total_Stops  Price  month  day  Airline  Source  \
0     18     2 50          0.0   3897      3   24        3       0   
1     84     7 25          2.0   7662      5    1        1       3   
2    118       19          2.0  13882      6    9        4       2   
3     91     5 25          1.0   6218      5   12        3       3   
4     29     4 45          1.0  13302      3    1        3       0   

   Destination  Dep_Time_Hours  Dep_Time_Minutes  Arrival_Time_hours  \
0            5              22                20                   1   
1            0               5                50                  13   
2            1               9                25                   4   
3            0              18                 5                  23   
4            5              16                50                  21   

   Arrival_Time_minutes Arrival_Time_day Arrival_Time_month Duration_hours  \
0                    10               22                Mar              2   
1                    15             None               None              7   
2                    25               10                Jun             19   
3                    30             None               None              5   
4                    35             None               None              4   

  Duration_minutes  
0               50  
1               25  
2             None  
3               25  
4               45

In [295]:
df.isna().sum()

Route                      0
Duration                   0
Total_Stops                1
Price                      0
month                      0
day                        0
Airline                    0
Source                     0
Destination                0
Dep_Time_Hours             0
Dep_Time_Minutes           0
Arrival_Time_hours         0
Arrival_Time_minutes       0
Arrival_Time_day        6348
Arrival_Time_month      6348
Duration_hours             0
Duration_minutes        1032
dtype: int64

In [296]:
df['Arrival_Time_day']=df["Arrival_Time_day"].fillna(0)


In [297]:
df["Arrival_Time_month"]=df["Arrival_Time_month"].fillna(value="Aa")
df["Duration_minutes"]=df["Duration_minutes"].fillna(0)

In [298]:
df.head()

Route Duration  Total_Stops  Price  month  day  Airline  Source  \
0     18     2 50          0.0   3897      3   24        3       0   
1     84     7 25          2.0   7662      5    1        1       3   
2    118       19          2.0  13882      6    9        4       2   
3     91     5 25          1.0   6218      5   12        3       3   
4     29     4 45          1.0  13302      3    1        3       0   

   Destination  Dep_Time_Hours  Dep_Time_Minutes  Arrival_Time_hours  \
0            5              22                20                   1   
1            0               5                50                  13   
2            1               9                25                   4   
3            0              18                 5                  23   
4            5              16                50                  21   

   Arrival_Time_minutes Arrival_Time_day Arrival_Time_month Duration_hours  \
0                    10               22                Mar              2   
1                    15                0                 Aa              7   
2                    25               10                Jun             19   
3                    30                0                 Aa              5   
4                    35                0                 Aa              4   

  Duration_minutes  
0               50  
1               25  
2                0  
3               25  
4               45

In [299]:
df=df.drop("Duration",axis=1)

In [300]:
df.dtypes

Route                     int32
Total_Stops             float64
Price                     int64
month                     int64
day                       int64
Airline                   int64
Source                    int64
Destination               int64
Dep_Time_Hours            int64
Dep_Time_Minutes          int64
Arrival_Time_hours        int64
Arrival_Time_minutes      int64
Arrival_Time_day         object
Arrival_Time_month       object
Duration_hours           object
Duration_minutes         object
dtype: object

In [301]:
df["Arrival_Time_month"].value_counts()

Aa     6348
May    1552
Jun    1423
Mar    1102
Apr     258
Name: Arrival_Time_month, dtype: int64

In [303]:
df["Arrival_Time_month"]=df["Arrival_Time_month"].str.replace("Aa","0")

In [308]:
df["Arrival_Time_month"]=df["Arrival_Time_month"].str.replace("Mar","3")
df["Arrival_Time_month"]=df["Arrival_Time_month"].str.replace("Apr","4")
df["Arrival_Time_month"]=df["Arrival_Time_month"].str.replace("May","5")
df["Arrival_Time_month"]=df["Arrival_Time_month"].str.replace("Jun","6")

In [309]:
df["Arrival_Time_month"].value_counts()

0    6348
5    1552
6    1423
3    1102
4     258
Name: Arrival_Time_month, dtype: int64

In [310]:
df["Arrival_Time_month"]=df["Arrival_Time_month"].astype(int)

In [311]:
df["Arrival_Time_day"]=df["Arrival_Time_day"].astype(int)

In [312]:
df["Duration_hours"]=df["Duration_hours"].astype(int)
df["Duration_minutes"]=df["Duration_minutes"].astype(int)

In [327]:
df.dtypes

Route                     int32
Total_Stops             float64
Price                     int64
month                     int64
day                       int64
Airline                   int64
Source                    int64
Destination               int64
Dep_Time_Hours            int64
Dep_Time_Minutes          int64
Arrival_Time_hours        int64
Arrival_Time_minutes      int64
Arrival_Time_day          int32
Arrival_Time_month        int32
Duration_hours            int32
Duration_minutes          int32
dtype: object

# DATA CLEANING : NULL VALUE IMPUTATION

In [329]:
df.isnull().sum()

Route                   0
Total_Stops             1
Price                   0
month                   0
day                     0
Airline                 0
Source                  0
Destination             0
Dep_Time_Hours          0
Dep_Time_Minutes        0
Arrival_Time_hours      0
Arrival_Time_minutes    0
Arrival_Time_day        0
Arrival_Time_month      0
Duration_hours          0
Duration_minutes        0
dtype: int64

In [330]:
df.dropna(inplace=True)

In [331]:
df.isna().sum()

Route                   0
Total_Stops             0
Price                   0
month                   0
day                     0
Airline                 0
Source                  0
Destination             0
Dep_Time_Hours          0
Dep_Time_Minutes        0
Arrival_Time_hours      0
Arrival_Time_minutes    0
Arrival_Time_day        0
Arrival_Time_month      0
Duration_hours          0
Duration_minutes        0
dtype: int64

In [332]:
df.head()

Route  Total_Stops  Price  month  day  Airline  Source  Destination  \
0     18          0.0   3897      3   24        3       0            5   
1     84          2.0   7662      5    1        1       3            0   
2    118          2.0  13882      6    9        4       2            1   
3     91          1.0   6218      5   12        3       3            0   
4     29          1.0  13302      3    1        3       0            5   

   Dep_Time_Hours  Dep_Time_Minutes  Arrival_Time_hours  Arrival_Time_minutes  \
0              22                20                   1                    10   
1               5                50                  13                    15   
2               9                25                   4                    25   
3              18                 5                  23                    30   
4              16                50                  21                    35   

   Arrival_Time_day  Arrival_Time_month  Duration_hours  Duration_minutes  
0                22                   3               2                50  
1                 0                   0               7                25  
2                10                   6              19                 0  
3                 0                   0               5                25  
4                 0                   0               4                45

# MODELLING

In [333]:
x=df.drop("Price",axis=1)
y=df["Price"]

In [334]:
x.shape

(10682, 15)

# LINEAR REGRESSION

In [335]:
from sklearn.model_selection import train_test_split

In [336]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, 
                                        test_size=0.3, random_state=0) 

In [337]:
from sklearn.linear_model import LinearRegression

In [338]:
lm=LinearRegression()

In [339]:
lm.fit(x_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [340]:
y_pred=lm.predict(x_test)

In [341]:
y_pred

array([9228.54089255, 8001.20851464, 4425.12488665, ..., 9486.82760964,
       9425.81232421, 8822.01000694])

In [342]:
lm.intercept_

9076.230530547276

In [343]:
list(zip(x_train.columns,lm.coef_))

[('Route', -11.898640142154767),
 ('Total_Stops', 4031.599460811193),
 ('month', -447.68135489695715),
 ('day', -73.18253060870994),
 ('Airline', 186.04469487610427),
 ('Source', -236.73795591494243),
 ('Destination', 53.92938594111833),
 ('Dep_Time_Hours', 49.822309090322946),
 ('Dep_Time_Minutes', -6.333821896433939),
 ('Arrival_Time_hours', -18.65727464210725),
 ('Arrival_Time_minutes', -9.614414597801908),
 ('Arrival_Time_day', -12.153115680533164),
 ('Arrival_Time_month', -39.92885329738813),
 ('Duration_hours', 83.67608361508054),
 ('Duration_minutes', -12.66954537401636)]

In [345]:
y_pred_train=lm.predict(x_train)

In [344]:
from sklearn.metrics import mean_squared_error,r2_score

In [346]:
mse=mean_squared_error(y_train, y_pred_train)
r_squared=r2_score(y_train,y_pred_train)

In [347]:
print("TRAINING ERROR AND R2 SCORE")
print("MSE : {}".format(mse))
print("R2 SCORE : {}".format(r_squared))

TRAINING ERROR AND R2 SCORE
MSE : 11748024.152735962
R2 SCORE : 0.44387585589095535


In [348]:
mse=mean_squared_error(y_test, y_pred)
r_squared=r2_score(y_test,y_pred)

In [349]:
print("TEST ERROR AND R2 SCORE")
print("MSE : {}".format(mse))
print("R2 SCORE : {}".format(r_squared))

TEST ERROR AND R2 SCORE
MSE : 12236658.362648156
R2 SCORE : 0.4332247502140262


In [350]:
from sklearn.metrics import mean_absolute_error

In [351]:
mean_absolute_error(y_test,y_pred)

2480.0522342735176

# LASSO AND RIDGE REGRESSION

In [354]:
import numpy as np

In [355]:
from sklearn.linear_model import Ridge,Lasso
from sklearn.model_selection import GridSearchCV

In [356]:
lambdas=np.linspace(1,100,100)

In [357]:
params={'alpha':lambdas}

In [358]:
model=Ridge(fit_intercept=True)

In [359]:
grid_search=GridSearchCV(model,param_grid=params,cv=10,scoring="neg_mean_absolute_error")

In [360]:
grid_search.fit(x_train,y_train)

GridSearchCV(cv=10, error_score='raise-deprecating',
             estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True,
                             max_iter=None, normalize=False, random_state=None,
                             solver='auto', tol=0.001),
             iid='warn', n_jobs=None,
             param_grid={'alpha': array([  1.,   2.,   3.,   4.,   5.,   6.,   7.,   8.,   9.,  10.,  11.,
        12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.,  20.,  21.,  22.,
        23.,  24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,  32.,...
        34.,  35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,  43.,  44.,
        45.,  46.,  47.,  48.,  49.,  50.,  51.,  52.,  53.,  54.,  55.,
        56.,  57.,  58.,  59.,  60.,  61.,  62.,  63.,  64.,  65.,  66.,
        67.,  68.,  69.,  70.,  71.,  72.,  73.,  74.,  75.,  76.,  77.,
        78.,  79.,  80.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,  88.,
        89.,  90.,  91.,  92.,  93.,  94.,  95.,  96.,  97.,  98.,  99.,
       100.]

In [362]:
grid_search.best_estimator_

Ridge(alpha=42.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [361]:
ridge_model=grid_search.best_estimator_

In [363]:
ridge_model.fit(x_train,y_train)

Ridge(alpha=42.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [364]:
list(zip(x_train.columns,ridge_model.coef_))

[('Route', -11.272307323395276),
 ('Total_Stops', 3902.3320768329854),
 ('month', -448.40067928500144),
 ('day', -73.43356369445408),
 ('Airline', 181.90165843849323),
 ('Source', -238.0382704136341),
 ('Destination', 52.069081774861736),
 ('Dep_Time_Hours', 48.60622642193634),
 ('Dep_Time_Minutes', -6.248770349317945),
 ('Arrival_Time_hours', -18.132204142959463),
 ('Arrival_Time_minutes', -9.721421686634839),
 ('Arrival_Time_day', -11.671979271645332),
 ('Arrival_Time_month', -38.37732957483775),
 ('Duration_hours', 89.67270690460009),
 ('Duration_minutes', -12.69679949024077)]

In [365]:
ytrainpred_ridge=ridge_model.predict(x_train)

In [366]:
ytrainpred_ridge

array([14001.27209146, 10541.10737064, 13762.76830521, ...,
       11962.19772611, 15303.01699746, 11340.83292886])

In [367]:
ypred_ridge=ridge_model.predict(x_test)

In [368]:
ypred_ridge

array([9181.49697016, 7977.75611201, 4486.61688231, ..., 9432.18383503,
       9407.04169892, 8792.21125116])

In [369]:
mse=mean_squared_error(y_test,ypred_ridge)
r2=r2_score(y_test,ypred_ridge)
r2_train=r2_score(y_train,ytrainpred_ridge)

In [370]:
print("Mean squared error : ",mse)
print("Train_r2_score : ",r2_train)
print("Test_r2_score : ",r2)

Mean squared error :  12240579.687951213
Train_r2_score :  0.4437417735060165
Test_r2_score :  0.4330431230032089


In [371]:
#lasso

In [372]:
lambdas=np.linspace(1,10,100)
model=Lasso(fit_intercept=True)
params={"alpha":lambdas}

In [373]:
grid_search1=GridSearchCV(model,param_grid=params,cv=10,scoring="neg_mean_absolute_error")

In [374]:
grid_search1.fit(x_train,y_train)

GridSearchCV(cv=10, error_score='raise-deprecating',
             estimator=Lasso(alpha=1.0, copy_X=True, fit_intercept=True,
                             max_iter=1000, normalize=False, positive=False,
                             precompute=False, random_state=None,
                             selection='cyclic', tol=0.0001, warm_start=False),
             iid='warn', n_jobs=None,
             param_grid={'alpha': array([ 1.        ,  1.09090909,  1.18181818,  1.27272727,  1.36363636,
        1.45454545,  1.5454...
        7.81818182,  7.90909091,  8.        ,  8.09090909,  8.18181818,
        8.27272727,  8.36363636,  8.45454545,  8.54545455,  8.63636364,
        8.72727273,  8.81818182,  8.90909091,  9.        ,  9.09090909,
        9.18181818,  9.27272727,  9.36363636,  9.45454545,  9.54545455,
        9.63636364,  9.72727273,  9.81818182,  9.90909091, 10.        ])},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='neg_mean_absolu

In [375]:
lasso_model=grid_search1.best_estimator_

In [376]:
lasso_model.fit(x_train,y_train)

Lasso(alpha=10.0, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [377]:
y_pred_lasso=lasso_model.predict(x_test)

In [378]:
mse=mean_squared_error(y_test,y_pred_lasso)
r2=r2_score(y_test,y_pred_lasso)


In [379]:
print("Mean squared error : ",mse)
print("r2_score : ",r2)

Mean squared error :  12241592.590525996
r2_score :  0.43299620757149393


# DECISION TREE REGRESSION

In [380]:
#decision tree regression
from sklearn.tree import DecisionTreeRegressor
DT = DecisionTreeRegressor(max_depth=15, min_samples_leaf=100)
DT.fit(x_train,y_train)


DecisionTreeRegressor(criterion='mse', max_depth=15, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=100,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')

In [381]:
ytrainpred_dt=DT.predict(x_train)

In [382]:
ytrainpred_dt

array([10447.06569343,  7866.91160221, 12880.4040404 , ...,
       13727.10857143, 14460.03496503, 13149.94339623])

In [383]:
ypred_dt=DT.predict(x_test)

In [384]:
mse=mean_squared_error(y_test,ypred_dt)
r2=r2_score(y_test,ypred_dt)
r2_train=r2_score(y_train,ytrainpred_dt)

In [385]:
print("Mean squared error : ",mse)
print("Train_r2_score : ",r2_train)
print("Test_r2_score : ",r2)

Mean squared error :  7782301.412638791
Train_r2_score :  0.6741422186746335
Test_r2_score :  0.6395408210037226


# RANDOM FOREST REGRESSION

In [386]:
from sklearn.ensemble import RandomForestRegressor

In [387]:
RF = RandomForestRegressor(max_depth=8, min_samples_leaf=150)

In [388]:
RF.fit(x_train,y_train)

C:\Users\RAMASAMY\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=8,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=150, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=10,
                      n_jobs=None, oob_score=False, random_state=None,
                      verbose=0, warm_start=False)

In [389]:
ytrainpred_RF=RF.predict(x_train)

In [392]:
ytrainpred_RF

array([11155.11004671, 13133.30227533, 13274.59912675, ...,
       12674.3909899 , 12638.29162374, 12104.31287138])

In [393]:
ypred_RF=RF.predict(x_test)

In [394]:
ypred_RF

array([12555.60715157,  6562.24059315,  5078.11771386, ...,
        5756.57933739, 11569.10996997,  6562.24059315])

In [395]:
from math import sqrt

In [396]:
mse=mean_squared_error(y_test,ypred_RF)
rmse=sqrt(mean_squared_error(y_test,ypred_RF))
r2=r2_score(y_test,ypred_RF)
r2_train=r2_score(y_train,ytrainpred_RF)

In [397]:
print("Mean squared error : ",mse)
print("Train_r2_score : ",r2_train)
print("Test_r2_score : ",r2)
print("rmse : ",rmse)

Mean squared error :  9521902.77588176
Train_r2_score :  0.5927726465675409
Test_r2_score :  0.5589662909351492
rmse :  3085.7580553053344


# FROM THE ABOVE MODELS WE CAN COME TO A CONCLUSION THAT , DECISION TREE MODEL GIVES THE BETTER ACCURACY OF 63% WITH MINIMIZED MSE. FURTHER TUNING OF THESE MODELS WITH RESPECT TO MULTI COLLINEARITY MAY RESULT IN HIGHER MODEL ACCURACY